### PR TITLE
Reduce guards of optimizer state dict to guard once per param group

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123413

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_dynamo/variables/optimizer.py [hunk 1]
```diff
@@ -179,17 +179,6 @@ def mark_static(x):
...
-        # NB: Its necessary to install the guards for optim.state first.
-        # optim.state is a dict with parameters as keys. Therefore, we just put
-        # ID_MATCH on parameters, instead of TENSOR_MATCH. When we install the
-        # guards for param_groups later, VariableTrackerCache just ensures that
-        # we directly return the cached tensor variable tracker without
-        # inserting the TENSOR_MATCH guard.
-        state_vt = LazyVariableTracker.realize_all(
-            VariableBuilder(tx, AttrSource(self.source, "state"))(self.value.state)
-        )
-
```
Comment (anijain2305) [inline torch/_dynamo/variables/optimizer.py:192]: Can you check that we are still doing DATA_PTR_MATCH guards on the parameters?

**Review by anijain2305 — APPROVED** [review]

File: torch/_dynamo/variables/optimizer.py (hunk 1 above)
Comment (mlazos) [reply torch/_dynamo/variables/optimizer.py:192]: Yeah we are

**Pull request closed without merging**